docs(cn): translate legacy/Component into Chinese (#1234)
reactjs/zh-hans.react.dev: src/content/reference/react/Component.md

<Pitfall>

> 我们建议将组件定义为函数而不是类。[了解如何迁移](#alternatives)

Reviewer: ```suggestion
我们建议将组件定义为函数而不是类。[了解如何迁移](#alternatives)。
```

Reviewer: 可以考虑调整为“我们建议使用函数式组件，而不是类式组件。”

> 我们建议使用函数组件，而不是类组件。[了解如何迁移](#alternatives)。

Reviewer: 统一使用「类式组件」与「函数式组件」。

</Pitfall>

<Intro>

> `Component` 是被定义为 [JavaScript 类](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) 的 React 基类，类组件仍然被 React 支持，但我们不建议在新代码中使用它们。

Reviewer 2: > `Component` is the base class for the React components defined as [JavaScript classes.] 
注意这里末尾有个句号，需要断句。

看翻译结果似乎没有注意到这个句号，所以导致结果有些偏差。

Reviewer: `Component` 是被定义为 JavaScript 类，此处表达有些拗口。

> `Component` 是一个 [JavaScript 类](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)，是被定义为react组件的基类，类组件仍然被 React 支持，但我们不建议在新代码中使用它们。

Reviewer: 注意，MDN 保留中文站点链接。

> `Component` 是一个 [JavaScript 类](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)，是被定义为react组件的基类，类组件仍然被 React 支持，但我们不建议在新代码中使用它们。

Reviewer: ```suggestion
`Component` 是一个 [JavaScript 类](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)，其被定义为 React 组件的基类，类式组件仍然被 React 支持，但我们不建议在新代码中使用它们。
```

---

## 参考 {/*reference*/}

### `Component` {/*component*/}

> 要想将 React 组件定义为一个类，你需要扩展内置的 `Component` 类并定义一个 [`render` 方法](#render)：

Reviewer: ```suggestion
如果想要使用类式组件，需要继承内置的 `Component` 类并定义 [`render` 方法](#render)：
```

}
```

只有 `render` 方法是必要的，其他方法是可选的。

> [查看以下的更多实例](#usage)

Reviewer: ```suggestion
[查看以下的更多实例](#usage)。
```

> [参加下方更多示例](#usage)。

Reviewer: ```suggestion
[参见下方更多示例](#usage)。
```

---

### `context` {/*context*/}

> 可以通过使用 `this.context` 来访问一个类式组件的 [context](/learn/passing-data-deeply-with-context)。只有当你使用 [`static contextType`](#static-contexttype) （最新的）或者 [`static contextTypes`](#static-contexttypes) （已被废弃的）来特别指定你想要接受 **哪一个** context 时它才会有效。

Reviewer: ```suggestion
类式组件可以通过使用 `this.context` 访问 [context](/learn/passing-data-deeply-with-context)。只有使用 [`static contextType`](#static-contexttype)（最新的）或 [`static contextTypes`](#static-contexttypes)（已废弃）特别指定想要接受 **哪一个** context 时才会有效。
```

> 一个类组件的 [context](/learn/passing-data-deeply-with-context) 可以通过使用 `this.context` 来实现。 只有当你使用 [`static contextType`](#static-contexttype) (更新的)或者 [`static contextTypes`](#static-contexttypes) (已被废弃) 来特别指定你想要接受 **哪一个** context 时它才会有效。

Reviewer: ```suggestion
一个类组件的 [context](/learn/passing-data-deeply-with-context) 可以通过使用 `this.context` 来实现。 只有当你使用 [`static contextType`](#static-contexttype)（更新的）或者 [`static contextTypes`](#static-contexttypes) (已被废弃) 来特别指定你想要接受 **哪一个** context 时它才会有效。
```

> 类式组件一次只能读取一个context。

Reviewer: ```suggestion
类式组件一次只能读取一个 context。
```

<Note>

> 在类式组件中读取 `this.props` 等同于在函数式组件中使用的 [声明式 props](/learn/passing-props-to-a-component#step-2-read-props-inside-the-child-component)。

Reviewer: ```suggestion
在类式组件中读取 `this.props` 等同于在函数式组件中使用 [声明式 props](/learn/passing-props-to-a-component#step-2-read-props-inside-the-child-component)。
```

> 在类组件中读取 `this.props` 等同于在函数式组件中使用的 [declaring props](/learn/passing-props-to-a-component#step-2-read-props-inside-the-child-component) 。

Reviewer 2: 这里的 declaring 不是关键字，可以考虑翻译为 声明式的 ？

<Deprecated>

这个 API 将在 React 未来的主版本中被移除。[使用 `createRef` 来代替](/reference/react/createRef)。

</Deprecated>

> 允许你获取此组件的 [非法字符串 ref](https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs) 

Reviewer: ```suggestion
允许你获取此组件的 [历史用法的字符串 ref](https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs)。
```

> 允许你获取此组件的 [过时的字符串 refs](https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs) 

Reviewer: ```suggestion
允许你获取此组件的 [过时的字符串 ref](https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs) 
```

> 允许你获取此组件的 [历史用法的字符串 ref](https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs)。

Reviewer 3: ```suggestion
允许你获取此组件 [过时的字符串 ref](https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs)。
```

---

### `state` {/*state*/}

> 使用 `this.state` 来访问一个类组件的 state 。 `state` 字段必须是一个对象。请不要直接改变 state 的值。如果你希望改变 state ，使用新的 state 来调用 `setState` 函数。

Reviewer 2: 这里的标点符号似乎有些多余的空格

<Note>

> 在类组件中定义 `state` 等同于在函数式组件中通过调用 [`useState`](/reference/react/useState) 函数所创造的 state 。

Reviewer 2: 格式有些问题，建议使用全局搜索: `[空格]。`，`[空格]，`等， 对这一系列规范进行检查。 
同时也要注意全局替换的影响范围，避免误伤了正确的内容。

  }
```

> 如果你使用更新的 JavaScript 语法的话，那么很少需要使用到 constructors 。相反，你可以使用现代浏览器和像 [Babel](https://babeljs.io/) 这样的工具都支持的[公有类字段语法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields)来重写上面的代码。

Reviewer 2: > ... 使用到 constructors 。

这里句号前不应该有空格

  }
```

> constructor 不应该包含任何副作用或者监听相关的代码。

Reviewer: ```suggestion
构造函数不应包含任何存在副作用或者事件监听相关的代码。
```

Reviewer: 此处原文使用「a constructor」，可以理解为「类中的构造器函数」，也可以直接当做 `constructor` 讲。

}
```

[查看更多示例](#adding-lifecycle-methods-to-a-class-component)。

#### 参数 {/*componentdidmount-parameters*/}

`componentDidMount` 不需要任何参数。

#### 返回值 {/*componentdidmount-returns*/}

`componentDidMount` 不应该返回任何值。

#### 说明 {/*componentdidmount-caveats*/}

- 当 [严格模式](/reference/react/StrictMode) 开启时，在开发环境中 React 会调用 `componentDidMount`，然后会立刻调用 [`componentWillUnmount`](#componentwillunmount)， 接着再次调用 `componentDidMount`。 这将帮助你注意到你是否忘记设置 `componentWillUnmount` 或者它的逻辑是否完全对应到 `componentDidMount` 的效果。

> - 虽然你可以在 `componentDidMount` 中立即调用 [`setState`](#setstate)，不过最好避免这样做。 因为这将触发一次额外的渲染，但是这是在浏览器更新屏幕之前发生的。在这种情况下即使 [`render`](#render) 被调用了两次，用户也无法看到中间的状态。请谨慎使用这种模式因为它可能会造成性能问题。在大多数情况下，你应该能在 [`constructor`](#constructor) 中设置初始的 state。 但是对于 modal 和 tooltip 等当你的渲染依赖于 DOM 节点的大小或位置情况下，这种方法可能是必要的。

Reviewer: ```suggestion
- 虽然你可以在 `componentDidMount` 中立即调用 [`setState`](#setstate)，不过最好避免这样做。因为这将触发一次额外的渲染，但是这是在浏览器更新屏幕之前发生的。在这种情况下即使 [`render`](#render) 被调用了两次，用户也无法看到中间的状态。请谨慎使用这种模式因为它可能会造成性能问题。在大多数情况下，你应该能在 [`constructor`](#constructor) 中设置初始的 state。但是对于 modal 和 tooltip 等当你的渲染依赖于 DOM 节点的大小或位置情况下，这种方法可能是必要的。
```

> - 虽然你可以在 `componentDidMount` 中立即调用[`setState`](#setstate)，不过最好避免这样做。 因为这将触发一次额外的渲染，但是这是在浏览器更新屏幕之前发生的。 在这种情况下即使 [`render`](#render) 被调用了两次， 用户也无法看到中间的状态。请谨慎使用这种模式因为它可能会造成性能问题。在大多数情况下， 你应该能在[`constructor`](#constructor)中设置初始的 state。 但是，对于 modals 和 tooltips 等当你的渲染依赖于 DOM 节点的大小或位置情况下，这种方法可能是必要的。

Reviewer: ```suggestion
- 虽然你可以在 `componentDidMount` 中立即调用[`setState`](#setstate)，不过最好避免这样做。 因为这将触发一次额外的渲染，但是这是在浏览器更新屏幕之前发生的。 在这种情况下即使 [`render`](#render) 被调用了两次， 用户也无法看到中间的状态。请谨慎使用这种模式因为它可能会造成性能问题。在大多数情况下， 你应该能在 [`constructor`](#constructor) 中设置初始的 state。 但是，对于 modal 和 tooltip 等当你的渲染依赖于 DOM 节点的大小或位置情况下，这种方法可能是必要的。
```

<Note>

> 对于大多数的使用场景来说，在类组件中一起定义 `componentDidMount`， `componentDidUpdate` 和 `componentWillUnmount` 等同于在函数组件中调用 [`useEffect`](/reference/react/useEffect)。在一些少数的情况，例如在浏览器绘制前执行代码很重要时，更像是等同于 [`useLayoutEffect`](/reference/react/useLayoutEffect)。

Reviewer: ```suggestion
对于大多数的使用场景来说，在类组件中一起定义 `componentDidMount`，`componentDidUpdate` 和 `componentWillUnmount` 等同于在函数组件中调用 [`useEffect`](/reference/react/useEffect)。在一些少数的情况，例如在浏览器绘制前执行代码很重要时，更像是等同于 [`useLayoutEffect`](/reference/react/useLayoutEffect)。
```

}
```

[查看更多示例](#adding-lifecycle-methods-to-a-class-component)。

#### 参数 {/*componentdidupdate-parameters*/}

> * `prevProps`：更新之前的 Props。`prevProps` 将会与 [`this.props`](#props) 进行比较来确定发生了什么改变。

Reviewer 3: ```suggestion
* `prevProps`：更新之前的 props。`prevProps` 将会与 [`this.props`](#props) 进行比较来确定发生了什么改变。
```

> * `prevState`：更新之前的 State。`prevState` 将会与 [`this.state`](#state) 进行比较来确定发生了什么改变。

Reviewer 3: ```suggestion
* `prevState`：更新之前的 state。`prevState` 将会与 [`this.state`](#state) 进行比较来确定发生了什么改变。
```

* `snapshot`: 如果你设置了 [`getSnapshotBeforeUpdate`](#getsnapshotbeforeupdate) 方法，那么 `snapshot` 将包含从该方法返回的值。否则它将是 `undefined`。

#### 返回值 {/*componentdidupdate-returns*/}

`componentDidUpdate` 不应该返回任何值。

> #### 说明 {/*componentdidupdate-caveats*/}

Reviewer: ```suggestion
#### 注意 {/*componentdidupdate-caveats*/}
```

> - 如果你定义了 [`shouldComponentUpdate`](#shouldcomponentupdate) 并且返回 false 的话，那么 `componentDidUpdate` 将不会被调用。

Reviewer: ```suggestion
- 如果你定义了 [`shouldComponentUpdate`](#shouldcomponentupdate) 并且返回 `false` 的话，那么 `componentDidUpdate` 将不会被调用。
```

- `componentDidUpdate` 内部的逻辑通常应该包含在比较 `this.props` 与 `prevProps` 以及 `this.state` 与 `prevState` 之中。否则就会存在创建无限循环的风险。

- 虽然可以在 `componentDidUpdate` 中立即调用 [`setState`](#setstate)，但最好尽可能避免这样做。它将触发一个额外的渲染，但它将在浏览器更新屏幕之前发生。这导致了即使 [`render`](#render) 在这种情况下会被调用两次，用户也看不到中间状态。这种模式通常会导致性能问题，但是对于 modal 和 tooltip 等当你的渲染依赖于 DOM 节点的大小或位置情况下，这种方法可能是必要的。

<Note>

> 对于大多数用例来说， 在类组件中一起定义 `componentDidMount`，`componentDidUpdate` 和 `componentWillUnmount` 等同于在函数组件中定义 [`useEffect`](/reference/react/useEffect)。在一些少数的情况，例如在浏览器绘制前执行代码很重要时，更像是等同于[`useLayoutEffect`](/reference/react/useLayoutEffect)。

Reviewer: ```suggestion
对于大多数用例来说，在类组件中一起定义 `componentDidMount`，`componentDidUpdate` 和 `componentWillUnmount` 等同于在函数组件中定义 [`useEffect`](/reference/react/useEffect)。在一些少数的情况，例如在浏览器绘制前执行代码很重要时，更像是等同于[`useLayoutEffect`](/reference/react/useLayoutEffect)。
```

<Deprecated>

此 API 已从 `componentWillReceiveProps` 重命名为 [`UNSAFE_componentWillReceiveProps`](#unsafe_componentwillreceiveprops)。旧名称已被弃用，在 React 未来主要的版本中，只有新名称才有效。

> 运行 [`rename-unsafe-lifecycles` 重构器](https://github.com/reactjs/react-codemod#rename-unsafe-lifecycles) 来自动更新你的组件。

Reviewer: ```suggestion
运行 [`rename-unsafe-lifecycles` codemod](https://github.com/reactjs/react-codemod#rename-unsafe-lifecycles) 来自动更新你的组件。
```

}
```

React 可能随时调用 `render`，因此你不应该假设它在特定时间运行。一般来说，`render` 方法应该返回一段 [JSX](/learn/writing-markup-with-jsx)，但也支持一些 [其他的返回类型](#render-returns)（如字符串）。为了计算返回的 JSX，`render` 方法可以读取 [`this.props`](#props)、[`this.state`](#state) 和 [`this.context`](#context)。

> 你应该将 `render` 方法编写为纯函数，这意味着如果 props、state 和 context 相同，它应该返回相同的结果。它也不应该包含副作用（例如设置监听）或与浏览器 API 交互。副作用应该发生在事件处理程序或 [`componentDidMount`](#componentdidmount) 等方法中。

Reviewer: ```suggestion
你应该将 `render` 方法编写为纯函数，这意味着如果 props、state 和 context 相同，它应该返回相同的结果。它也不应该包含副作用（例如订阅监听事件）或与浏览器 API 交互。副作用应该发生在事件处理程序或 [`componentDidMount`](#componentdidmount) 等方法中。
```
